internal/worker/firewaller/firewaller.go

```diff
@@ -1501,171 +1518,279 @@ func (ad *applicationData) scopedContext() (context.Context, context.CancelFunc)
 
-type remoteRelationData struct {
-	catacomb      catacomb.Catacomb
-	fw            *Firewaller
-	relationReady chan remoteRelationInfo
+	// For consuming side, we need to identify which endpoint is local and which
+	// is remote.
+	if len(rel.Endpoints) != 2 {
+		return errors.Errorf("relation %v has %d endpoints, expected 2", tag.Id(), len(rel.Endpoints))
+	}
 
-	tag                 names.RelationTag
-	localApplicationTag names.ApplicationTag
-	relationToken       string
-	remoteModelUUID     string
-	endpointRole        charm.RelationRole
-	isOffer             bool
+	// Try to identify the remote application by querying both endpoints.
+	var localEndpoint domainrelation.Endpoint
+	var remoteApplicationName string
+	var remoteModelUUID string
 
-	crossModelFirewallerFacade CrossModelFirewallerFacadeCloser
+	for _, ep := range rel.Endpoints {
+		remoteApps, err := fw.crossModelRelationService.RemoteApplications(ctx, []string{ep.ApplicationName})
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if len(remoteApps) > 0 && remoteApps[0].Result != nil {
+			// This is the remote application.
+			remoteApplicationName = ep.ApplicationName
+			remoteModelUUID = remoteApps[0].Result.ModelUUID
+		} else {
+			// This is the local application.
+			localEndpoint = ep
+		}
+	}
 
-	// These values are updated when ingress information on the
-	// relation changes in the model.
-	ingressRequired bool
-	networks        set.Strings
+	if remoteApplicationName == "" {
+		return errors.Errorf("could not identify remote application for relation %v", tag.Id())
+	}
+
+	data := &remoteRelationData{
+		fw:                  fw,
+		tag:                 tag,
+		localApplicationTag: names.NewApplicationTag(localEndpoint.ApplicationName),
+		relationUUID:        rel.UUID,
+		remoteModelUUID:     remoteModelUUID,
+	}
+	var watchEgressLoop func() error
+	if localEndpoint.Role == charm.RoleRequirer {
+		// Local endpoint is requirer, so watch local egress and publish to
+		// remote (offering model).
+		watchEgressLoop = data.watchLocalEgressPublishRemote
+	} else {
+		// Local endpoint is provider, so watch remote (offering model) egress
+		// and apply locally.
+		watchEgressLoop = data.watchRemoteEgressApplyLocal
+	}
+
+	// Start the worker which watches for ingress/egress address changes
+	if err := fw.relationWorkerRunner.StartWorker(ctx, tag.Id(), func(ctx context.Context) (worker.Worker, error) {
+		// This may be a restart after an api error, so ensure any previous
+		// worker is killed and the catacomb is reset.
+		data.Kill()
+		data.catacomb = catacomb.Catacomb{}
+
+		if err := catacomb.Invoke(catacomb.Plan{
+			Name: "firewaller-consumer-relation",
+			Site: &data.catacomb,
+			Work: watchEgressLoop,
+		}); err != nil {
+			return nil, errors.Trace(err)
+		}
+		return data, nil
+	}); err != nil {
+		return errors.Annotate(err, "error starting consumer relation worker")
+	}
+	fw.relationIngress[tag] = data
```

Comment:
This is dumb. We should be pushing the information into the type, not keeping another independent map hanging around.

Reply:
These two comments will require a bit more work, and will be done in a follow-up patch as decided.